README.md
# qtlPlots

### Install
```r
devtools::install_github("GabrielHoffman/qtlPlots")
```

### Description
Plotting functions take in a `GRanges` object where `scores` defines the value to be plotted on the y-axis:
 
 * `plotMht()` Manhattan plot for local region
 * `plotPosterior()` Posterior probabilities in local region

 In addition, there are 2 other functions:
 
 * `plotEnsGenes()` Plot ENSEMBL genes
 * `plotBed()` Plot genome intervals


Here is a simple example for simulated data:
 
```r
suppressPackageStartupMessages({
library(GenomicRanges)
library(qtlPlots)
})
 
# simple example of QTL results
df = data.frame(Chr         = 1,
                Position    = 1:100, 
                Variant     = paste0('SNP', 1:100),
                p.value     = runif(100))

# Convert to GRanges
gr = with(df,   GRanges(Chr, 
                IRanges(start   = Position, 
                        width   = 1, 
                        Variant = Variant, 
                        score   = -log10(p.value))))
 
# where: genome interval to show
wh = GRanges(1, IRanges(1, 100))			
 						
# example plot
plotMht(gr, wh, recombRate=FALSE) 
```
 
# Example on real caQTL data 

You can also add any other plots to this for eQTL and GWAS

```r
# test combined plotting
suppressPackageStartupMessages({
library(qtlPlots)
library(ggplot2)
library(GenomicRanges) 
})

ensdb = EnsDb.Hsapiens.v75::EnsDb.Hsapiens.v75 # genes on hg19
# ensdb = EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86 # genes on hg38

# read data
dir <- system.file("data", package="qtlPlots")	
gr_caQTL = readRDS(paste0(dir, '/', "gr_caQTL.RDS"))
gr_peaks = readRDS(paste0(dir, '/', "gr_peaks.RDS"))
gr_finemap = readRDS(paste0(dir, '/', "gr_finemap.RDS"))

# plot results for this peak
targetPeak = 'Peak_82668'

# where: get interval to plot
wh = range(gr_caQTL[gr_caQTL$Gene == targetPeak])

# only keep results for target peak
gr_caQTL_sub = gr_caQTL[which(gr_caQTL$Gene == targetPeak)]
gr_finemap = gr_finemap[gr_finemap$feature == targetPeak]

# Manhattan plot
################

# identify variants in the candidate set
gr_caQTL_sub$inCandidateSet = FALSE
idx = gr_caQTL_sub$Variant %in% gr_finemap$Variant
gr_caQTL_sub$inCandidateSet[idx] = TRUE

# recomb rate is hg19
fig_mht = plotMht( gr_caQTL_sub, wh )

# Posteriors
#############

fig_post = plotPosterior( gr_finemap, wh )

# Plot genes
############

fig_gene = plotEnsGenes( ensdb, wh)

# ATAC-seq peaks
#################

# highlight the target peak
gr_peaks$status = 0
gr_peaks$status[names(gr_peaks) == targetPeak] = 1

fig_bed = plotBed( gr_peaks, wh )

# Combine into multi-track plot
###############################

fig_track = ggbio::tracks(
     "caQTL"        = fig_mht, 
    "caQTL Fine map"= fig_post,
    'Gene'          = fig_gene,
    'ATAC'          = fig_bed,
    xlim = wh,
    padding = unit(-.65, "lines"), 
    label.bg.fill="navy", 
    label.text.color="white",
    heights=c(1, .6, .2, .1),
    title = targetPeak) 
fig_track@mutable['Genes'] = FALSE
fig_track
```

 


NEWS.md
# qtlPlots 0.0.7
 - March 3, 2023
 - fix issues from `R CMD check`


R/plotBed.R

# Gabriel Hoffman
# July 22, 2021
# based on mmQTL_plots code 

#' Plot BED intervals
#'
#' Plot BED intervals
#'
#' @param grObj GRanges object with scores as the posterior inclusion probability
#' @param wh genome interval to plot
#'
#' @import ggplot2 GenomicRanges 
#' @importFrom scales comma
#' @importFrom IRanges `%within%`
#' @export
plotBed = function( grObj, wh ){

    gr_sub = grObj[grObj %within% wh]

    # convert to data.frame
    df_loc = as.data.frame(gr_sub)

    # Pass R CMd check 
    status = NULL

    if( ! ("status" %in% colnames(df_loc)) ){
        df_loc$status = 0
    }

    df_loc$status = factor(df_loc$status)

    col = c()
    if( any(df_loc$status==0)){
        col = append(col, "black")
    }
    if( any(df_loc$status==1)){
        col = append(col, "red")
    }

    df_loc$start = pmax(df_loc$start, start(wh))
    df_loc$end = pmin(df_loc$end, end(wh))

    ggplot(df_loc) + geom_segment(aes(x=start, y=1, xend = end, yend=1, color=status), size=4) + scale_x_continuous(expand=c(0,0), limits=c(start(wh), end(wh)), labels=comma) + scale_y_continuous(expand=c(0,0), limits=c(0,1)) + scale_color_manual(values = col) + theme_bw() + theme(legend.position = "none", axis.title=element_blank(), axis.ticks.y=element_blank(), axis.line.y=element_blank(), panel.grid=element_blank(), axis.text.y=element_blank(),panel.background = element_blank(), strip.background = element_blank(), rect = element_rect(fill="white", linetype=0))
}





R/plotMht.R
# Gabriel Hoffman
# July 22, 2021
# based on mmQTL_plots code 

#' Zoom in manhattan plot
#'
#' Zoom in manhattan plot
#'
#' @param grObj GRanges object with scores as the -log10 p-value
#' @param wh genome interval to plot
#' @param size font size
#' @param ptSize size for points
#' @param recombRate show recombination rate
#'
#' @import ggplot2 GenomicRanges GenomeInfoDb
#' @importFrom scales comma
#' @importFrom stats approx
#' @importFrom IRanges `%within%`
#' @export
plotMht = function( grObj, wh, size=8, ptSize = 1, recombRate = TRUE ){

	if( length(grObj) == 0){
		fig = ggplot() + theme_void() + scale_x_continuous(expand=c(0,0), labels=comma, limits=c(start(wh), end(wh)))
		return(fig)
	}


    # Pass R CMD check
    inCandidateSet = rate = NULL

	# load recombination rate
	dir <- system.file("data", package="qtlPlots")
	# dir = "/Users/gabrielhoffman/workspace/repos/qtlPlots/data/"
    gr_recomb <- readRDS(file.path(dir,"gr_recomb_hg19.RDS"))

	seqlevelsStyle(grObj) = "UCSC"
	seqlevelsStyle(wh) = "UCSC"
	seqlevelsStyle(gr_recomb) = "UCSC"

    # keep only entries on the same chromosome as grObj 
    # gr_recomb = gr_recomb[seqnames(gr_recomb) %in% seqnames(wh)] 
    # gr_recomb = gr_recomb[S4Vectors::match(seqnames(gr_recomb), seqnames(wh))]
    # gr_recomb = subset(gr_recomb, array(seqnames(gr_recomb)==unique(seqnames(wh))))
    gr_recomb = gr_recomb[as.character(seqnames(gr_recomb)) %in% as.character(seqnames(wh))]

    # extact scores *within* the region wh
    gr_sub = grObj[grObj %within% wh]

    if( is.null(gr_sub$inCandidateSet) ){
    	gr_sub$inCandidateSet = FALSE
    }

    df_sub = data.frame(Position=start(gr_sub), score=score(gr_sub), inCandidateSet=gr_sub$inCandidateSet)

    if( recombRate ){
	    # recombination rate
		# expand window until a recombination measurement is found 
		wh_query = wh
		n = 0
		while( n < 1 ){
			# get intersection
			idx = which(gr_recomb %within% wh)

			# count number
			n = length(idx)

			# expand interval
			wh_query = flank(wh_query, width=width(wh_query)*2, both=TRUE)
		}

		# then expand one more step in each direction to avoid edge effects
		# but avoid steping over the last element
		idx_augment = c(min(idx)-1, idx, max(idx)+1)
		idx_augment = pmin(pmax(1, idx_augment), length(gr_recomb))
		gr_recomb_sub = gr_recomb[idx_augment] 

		# Add recombination rate to first plot
		# linear interpolation evaluate at points in resComposite
		app = approx( start(gr_recomb_sub), gr_recomb_sub$recomb_rate, df_sub$Position )
		df_sub$rate = pmin(app$y, 100) # cap local rate at 100
		# df_window$log10.p.value = pmin(300,df_window$log10.p.value)
	}

	ymax = max(df_sub$score)
	ymax.rate = ymax

	# if( (ord == 1) || ! showConditional ){
		fig = ggplot(df_sub, aes(Position, score, color=inCandidateSet)) + scale_color_manual(values = c("FALSE"="black", "TRUE"="red")) + theme_classic(size)
	# }else{		
	# 	df_window[,cat := paste(inCandSet, eQTL_order < ord)]
	# 	df_window$cat = factor(df_window$cat, c("no TRUE", "no FALSE", "yes FALSE"))
	# 	setorder(df_window, eQTL_order, cat)
	# 	fig_eqtl = ggplot(df_window, aes(Position, log10.p.value, color=cat)) + scale_color_manual(values = c("grey", "black", "red"))
	# }

	fig = fig + geom_point(size=ptSize) + ylab(bquote(-log[10]~P)) + scale_x_continuous(expand=c(0,0), labels=comma, limits=c(start(wh), end(wh))) 

	if( recombRate ){
		fig = fig + scale_y_continuous(expand=c(0,0), limits=c(0, ymax*1.05), sec.axis = sec_axis(~./(ymax.rate/100), name = "Recombination\nrate [cM/Mb]")) + geom_line(aes(y=rate*(ymax.rate/100)), color="dodgerblue", size=.5)
	}

	fig + theme(legend.position="none")
}











R/plot_mht.R


#' Zoom in manhattan plot
#'
#' Zoom in manhattan plot
#'
#' @param grObj GRanges object with scores as the -log10 p-value
#' @param wh genome interval to plot
#'
#' @import ggplot2 GenomicRanges IRanges
#' @importFrom scales comma
#' @importFrom stats approx
#' @export
plot_mht = function( grObj, wh ){

	# load recombination rate
#	dir <- system.file("extdata", package="qtlPlots")
	dir = "/Users/gabrielhoffman/workspace/repos/qtlPlots/data/"
    gr_recomb <- readRDS(file.path(dir,"gr_recomb_hg19.RDS"))

    # keep only entries on the same chromosome as grObj 
    # gr_recomb = gr_recomb[seqnames(gr_recomb) %in% unique(seqnames(wh))] 
    gr_recomb = subset(gr_recomb, seqnames==unique(seqnames(wh)))

    # extact scores *within* the region wh
    gr_sub = grObj[grObj %within% wh]
    df_sub = with(gr_sub, data.frame(Position=start, score=score, inCandidateSet=inCandidateSet))

    # recombination rate
	# expand window until a recombination measurement is found 
	wh_query = wh
	n = 0
	while( n ==0 ){
		idx = which(gr_recomb %within% wh)

		n = length(idx)
		wh_query = flank(wh_query, width=width(wh_query)*2, both=TRUE)
	}

	# then expand one more step in each direction to avoid edge effects
	# but avoid steping over the last element
	idx_augment = c(min(idx)-1, idx, max(idx)+1)
	idx_augment = pmin(pmax(1, idx_augment), length(gr_recomb))
	gr_recomb_sub = gr_recomb[idx_augment] 

	# Add recombination rate to first plot
	# linear interpolation evaluate at points in resComposite
	app = approx( start(gr_recomb_sub), gr_recomb_sub$recomb_rate, df_sub$Position )
	df_sub$rate = pmin(app$y, 100) # cap local rate at 100
	# df_window$log10.p.value = pmin(300,df_window$log10.p.value)

	ymax = max(df_sub$score)
	ymax.rate = ymax

	# if( (ord == 1) || ! showConditional ){
		fig = ggplot(df_sub, aes(Position, score, color=inCandidateSet)) + scale_color_manual(values = c("black", "red")) + theme_classic()
	# }else{		
	# 	df_window[,cat := paste(inCandSet, eQTL_order < ord)]
	# 	df_window$cat = factor(df_window$cat, c("no TRUE", "no FALSE", "yes FALSE"))
	# 	setorder(df_window, eQTL_order, cat)
	# 	fig_eqtl = ggplot(df_window, aes(Position, log10.p.value, color=cat)) + scale_color_manual(values = c("grey", "black", "red"))
	# }

	fig = fig + geom_point(size=.4) + ylab(bquote(-log[10]~P)) + scale_x_continuous(expand=c(0,0), label=comma) + scale_y_continuous(expand=c(0,0), limits=c(0, ymax*1.05), sec.axis = sec_axis(~./(ymax.rate/100), name = "Recombination\nrate [cM/Mb]")) + geom_line(aes(y=rate*(ymax.rate/100)), color="dodgerblue", size=.5) + theme(legend.position="none")

	fig
}

R/plot_genes.R
# Gabriel Hoffman
#
# October 2, 2020
# Adapt decorate::plotEnsGenes to plot using ggplot2 instead of grid

# library(scales)

# adapted from decorate::get_exon_coords(), but return gene_id here
get_exon_coords = function (ensdb, query, biotypes = c("protein_coding")){
    if (!is(ensdb, "EnsDb")) {
        stop("ensdb must be an ENSEMBL databsed of type EnsDb")
    }
    gr_exon = exonsByOverlaps(ensdb, query, columns = c("exon_seq_start", 
        "exon_seq_end", "symbol", "gene_biotype", "tx_seq_start", 
        "tx_seq_end", "tx_cds_seq_start", "tx_cds_seq_end", 'gene_id'))
    if (!is.na(biotypes)) {
        gr_exon = gr_exon[gr_exon$gene_biotype %in% biotypes]
    }
    gr_exon
}


#' Plot ENSEMBL genes
#' 
#' Plot ENSEMBL genes in region
#' 
#' @param ensdb ENSEMBL database object like EnsDb.Hsapiens.v86 
#' @param wh genome interval to plot
# @param plot_lines_distance veritcal distance between genes
# @param vp viewport
#' @param splice_variants if TRUE, show multiple transcripts from the same gene
#' @param non_coding if TRUE, also show non-coding genes
#' @param arrow.size arrow size for genes
#' @param ensGene show specific non-coding locus
#'
#' @return ggplot2 of genome region
#' 
#' @examples
#' library(EnsDb.Hsapiens.v86)
#' library(GenomicRanges) 
# library(grid) 
#' 
#' # gene database
#' ensdb = EnsDb.Hsapiens.v86
#' 
#' # interval
#' query = GRanges("20", IRanges(62045027,62164563))
#'
#' # plot genes
#' plotEnsGenes( ensdb, query)
#' 
#' @export
#' @import GenomicRanges ggplot2
#' @importFrom scales comma
# @import grid
#' @importFrom data.table data.table
# @importFrom grDevices pdf dev.off
plotEnsGenes = function(ensdb, wh, splice_variants = FALSE, non_coding = FALSE, arrow.size=0.05, ensGene = NULL){

    minRange = start(wh)
    maxRange = end(wh)
    chromosome = seqnames(wh)

	gr = GRanges(gsub( "^chr", "", chromosome), IRanges(minRange, maxRange))

	if( non_coding ){
		gr_exons = get_exon_coords( ensdb, gr, NA )
	}else{
		gr_exons = decorate::get_exon_coords( ensdb, gr, 'protein_coding' )

		# get locus for ensGene (even if it is non-coding)
		if( !is.null(ensGene) ){
			# get all loci
			gr_exons_focus = get_exon_coords( ensdb, gr, NA )

			# add locus of focus to other loci
			gr_exons_focus = gr_exons_focus[gr_exons_focus$gene_id == ensGene]
			gr_exons = unique(c(gr_exons, gr_exons_focus))
		}
	}

    # get gene coordinates
    if( non_coding ){
        biotype = NA
    }else{
        biotype = c("protein_coding")
    }

    if( length(gr_exons) > 0){
   
    	df = data.table(data.frame(gr_exons))

        gene_biotype = 0
        symbol = 0
        tx_cds_seq_end = 0
        tx_cds_seq_start = 0
        tx_seq_end = 0
        tx_seq_start = 0

    	if( !splice_variants ){
    		# single body per gene
    		suppressWarnings(df_wide <- df[,data.frame(
    			gene_name = unique(symbol),
    			chrom = unique(seqnames),
    			strand = unique(strand), 
    			exonStarts = paste(start, collapse=','), 
    			exonEnds = paste(end, collapse=','),
    			exonCount = length(start),
    			biotype = unique(gene_biotype),
    			txStart = min(tx_seq_start, na.rm=TRUE),
    			txEnd = max(tx_seq_end, na.rm=TRUE),
    			cdsStart = as.integer(min(tx_cds_seq_start, na.rm=TRUE)),
    			cdsEnd = as.integer(max(tx_cds_seq_end, na.rm=TRUE)),
    			stringsAsFactors=FALSE),by=c("symbol")])
    	}else{
    		stop("splice_variants = TRUE not currently supported ")
    		# multiple transcripts per gene		
            suppressWarnings(df_wide <- df[,data.frame(
    			gene_name = unique(symbol),
    			chrom = unique(seqnames),
    			strand = unique(strand), 
    			exonStarts = paste(start, collapse=','), 
    			exonEnds = paste(end, collapse=','),
    			exonCount = length(start),
    			biotype = unique(gene_biotype),
    			txStart = tx_seq_start,
    			txEnd = tx_seq_end,
                cdsStart = as.integer(min(tx_cds_seq_start, na.rm=TRUE)),
                cdsEnd = as.integer(max(tx_cds_seq_end, na.rm=TRUE)),
    			stringsAsFactors=FALSE),by=c("symbol", 'tx_seq_start', 'tx_seq_end')])
    	}
        t = data.frame(df_wide, stringsAsFactors=FALSE)
        t$plot_line <- 0
        t$plot_line[1] <- 1
    }else{
        t = data.frame()
    }
	
	# get min and max values based on strand
	t$txStart = pmax(t$txStart, minRange)
	t$txEnd = pmin(t$txEnd, maxRange)

	t$txMin = ifelse( t$strand == "+", t$txStart, t$txEnd )
	t$txMax = ifelse( t$strand == "+", t$txEnd, t$txStart )

	plot_lines_no <- 1

    if (dim(t)[1] > 1) {
        for(i in 2:dim(t)[1]){
            # gene_name_width <- Range/map_len * convertWidth(grobWidth(textGrob(paste(t[1,"gene_name"], "   "), gp = gpar(fontsize = 7))),
                # "npc", valueOnly = TRUE)
            gene_name_width = width(gr) / 10
            for (j in 1:plot_lines_no) {
                if (max(t[1:(i - 1), ][t[, "plot_line"] == j, "txEnd"], na.rm = TRUE) < t[i, "txStart"] - gene_name_width*2) {
                  t[i, "plot_line"] <- j
                  break
                }
            }
            if (!t[i, "plot_line"])
                t[i, "plot_line"] <- plot_lines_no <- plot_lines_no + 1
        }

       	t$plot_line = max(t$plot_line) + 1 - t$plot_line
    }   

    if( nrow(t) > 0 ){

		# genes using ggplot2
		ylim = c(min(t$plot_line), max(t$plot_line))
		d = ylim[2] - ylim[1]

		fig = ggplot(t, aes(x=txMin, xend=txMax, y=plot_line, yend=plot_line, color = ifelse(biotype=='protein_coding', '1', '2'), label=paste0('  ',symbol, '  '), hjust=ifelse(strand=='+', 1, 0))) + geom_segment( arrow = arrow(length = unit(arrow.size, "npc"), end="last",type = "open")) + ylim(ylim[1] - 0.3*d, ylim[2] + 0.1*d) + geom_text(size=2.2) + scale_x_continuous(label=comma, expand=c(0,0), limits=c(minRange, maxRange)) + scale_color_manual(values=c("navy", "grey40")) 
	}else{
		fig = ggplot() + scale_x_continuous(label=comma, expand=c(0,0), limits=c(minRange, maxRange))
	}

	fig + theme_bw() + theme(legend.position = "none", axis.title=element_blank(), axis.ticks.y=element_blank(), axis.line.y=element_blank(), panel.grid=element_blank(), axis.text.y=element_blank(),panel.background = element_blank(), strip.background = element_blank(), rect = element_rect(fill="white", linetype=0))
}

# chr20 47223127-48223127
# chromosome = "9"
# minRange = 19000000
# maxRange = 19300000


# ,panel.border = element_blank(), panel.grid.major = element_blank(),
# 		panel.grid.minor = element_blank(), axis.line = element_blank(),
# 		axis.title=element_blank(),
#         axis.text=element_blank(),
#         axis.ticks=element_blank())




R/plotPosterior.R
# Gabriel Hoffman
# July 22, 2021
# based on mmQTL_plots code 


#' Zoom in manhattan plot of posteriors
#'
#' Zoom in manhattan plot of posteriors
#'
#' @param grObj GRanges object with scores as the posterior inclusion probability
#' @param wh genome interval to plot
#' @param size font size
#' @param ymax max value on y axis
#'
#' @import ggplot2 GenomicRanges 
#' @importFrom scales comma
#' @importFrom IRanges `%within%`
#' @export
plotPosterior = function( grObj, wh, size=8, ymax=1 ){

	# extact scores *within* the region wh
    gr_sub = grObj[grObj %within% wh]
    df_sub = data.frame(Position=start(gr_sub), score=score(gr_sub))

	fig = ggplot(df_sub, aes(Position, score)) + theme_classic(size)

	fig + geom_point(size=.4, color="red") + ylab("Posterior") + scale_y_continuous(expand=c(0,0.01), limits=c(0, ymax)) + scale_x_continuous(expand=c(0,0), limits=c(start(wh), end(wh)), labels=comma) + theme(legend.position="none") 
}	


R/make_plots.R
# Gabriel Hoffman
#
# October 2, 2020
# Intersect mmQTL with CAUSALdb

# library(RColorBrewer)
# library(EnsDb.Hsapiens.v75)

# Plot results
##############

make_plot = function( ensGene, window = 5e5, ord=1, non_coding=FALSE, showConditional=FALSE ){

	# geneSymbol = tryCatch(
	# 		mapIds(org.Hs.eg.db,
	#                      keys=ensGene,
	#                      column="SYMBOL",
	#                      keytype="ENSEMBL")
	# 		, error = function(e) '')

	# get gene name from ENSEMBL id
	geneInfo = select(EnsDb.Hsapiens.v75, keys=ensGene, keytype="GENEID", column=c('GENENAME', 'GENEBIOTYPE'))
	geneSymbol = geneInfo$GENENAME

	# get location of top fine-mapped eQTL SNP	
	df_fm = df_finemap[(Gene == ensGene) & (eQTL_order == ord),]
	pos_center = df_fm[which.max(PIP),Position]
	wh = GRanges(paste0("chr", df_fm$Chr[1]), IRanges(pos_center - window, pos_center + window))

	# eQTL
	if( (ord == 1) || ! showConditional ){		
		df_window = df_eqtl[(Gene == ensGene) & (eQTL_order == ord),]
	}else{		
		df_window = df_eqtl[(Gene == ensGene) & (eQTL_order <= ord),]
	}

	# get candidate causal set
	df_window = df_window[(Variant %in% df_fm$Variant) & (eQTL_order == ord),inCandSet := "yes"]
	df_window$inCandSet[is.na(df_window$inCandSet)]  = 'no'

	# recombination rate
	# expand window until a recombination measurement is found 
	scl = 1
	n = 0
	while( n ==0 ){
		df_recomb_sub = df_recomb[(position > start(wh)/scl) & (position < end(wh)*scl) & (chromosome == df_window$Chr[1]),]
		n = nrow(df_recomb_sub)
		scl = scl * 3
	}

	# Add recombination rate to first plot
	# linear interpolation evaluate at points in resComposite
	app = approx( df_recomb_sub$position, df_recomb_sub$recomb_rate, df_window$Position )
	df_window$rate = pmin(app$y, 100) # cap local rate at 100
	df_window$log10.p.value = pmin(300,df_window$log10.p.value)

	ymax = max(df_window$log10.p.value)
	ymax.rate = ymax
	
	if( (ord == 1) || ! showConditional ){
		fig_eqtl = ggplot(df_window, aes(Position, log10.p.value, color=inCandSet)) + scale_color_manual(values = c("black", "red"))
	}else{		
		df_window[,cat := paste(inCandSet, eQTL_order < ord)]
		df_window$cat = factor(df_window$cat, c("no TRUE", "no FALSE", "yes FALSE"))
		setorder(df_window, eQTL_order, cat)
		fig_eqtl = ggplot(df_window, aes(Position, log10.p.value, color=cat)) + scale_color_manual(values = c("grey", "black", "red"))
	}

	fig_eqtl = fig_eqtl + geom_point(size=.4) + ggtitle("FINEMAP - gene") + ylab(bquote(-log[10]~P)) + scale_x_continuous(expand=c(0,0)) + scale_y_continuous(expand=c(0,0), limits=c(0, ymax*1.05), sec.axis = sec_axis(~./(ymax.rate/100), name = "Recombination\nrate [cM/Mb]")) + geom_line(aes(y=rate*(ymax.rate/100)), color="dodgerblue", size=.5) + theme(legend.position="none")

	# eQTL fine mapping
	# df_fm = df_finemap[(Gene == ensGene) & (eQTL_order == ord),]
	#, color=as.character(eQTL_order)
	fig_finemap_gene = ggplot(df_fm, aes(Position, PIP, size=PIP^2)) + geom_point(color="red") + ggtitle("Finemap - gene") + ylab("Posterior") + ylim(0,1) + scale_x_continuous(expand=c(0,0)) + scale_size_continuous(limits = c(0, 1), range(0.1, 6)) 

	# CAUSALdb
	# c('OT294', 'OT344', 'OT311', 'GA553', 'OT310')
	df_pip = get_credible_set( causalDB, unique(df_show$meta_id), wh)
	setorder(df_pip, Trait, -FINEMAP)

	# intersect finemapping from gene (df_fm) and GWAS (df_pip)
	df_coloc = merge(df_fm, df_pip, by.x="Position", by.y="BP")
	df_select = df_coloc[, data.frame(max=max(FINEMAP*PIP)),by='meta_id']

	# only keep GWAS that shares a causal variant with eQTl at least 0.01 
	df_coloc = df_coloc[meta_id %in% df_select[max>0.01,meta_id],]

	# keep GWAS fine mapping for these traits
	df_pip = df_pip[(meta_id %in% df_coloc$meta_id) | (Trait %in% df_coloc$Trait),]
	df_pip$Trait = factor(df_pip$Trait)

	if( nrow(df_pip) > 0){

		if( nlevels(df_pip$Trait) < 8 ){
			cols = brewer.pal(n = nlevels(df_pip$Trait), name = "Dark2")
		}else{
			cols = rainbow( nlevels(df_pip$Trait) )
		}

		fig_causal = ggplot(df_pip, aes(BP, FINEMAP, color=Trait, size=FINEMAP^2)) + geom_point() + ylim(0,1) + scale_x_continuous(expand=c(0,0)) + ylab("Posterior") + scale_size_continuous(limits = c(0, 1), range(0.1, 6)) + scale_color_manual(values = cols) # + geom_text_repel(data=subset(df_pip, !duplicated(Trait)), aes(label=Trait), size=3, force=100, nudge_x = .1, nudge_x=window/2)
	}else{
		fig_causal = ggplot() + scale_x_continuous(expand=c(0,0), limits=c(start(wh), end(wh))) + ylim(0,1) + ylab("Posterior")
	}

	# plot coloc
	# df_coloc = merge(df_fm, df_pip, by.x="Position", by.y="BP")
	df_coloc = df_coloc[FINEMAP*PIP>0,]
	if( nrow(df_coloc) > 0){
		df_coloc$Trait = factor(df_coloc$Trait, levels(df_pip$Trait))
		df_coloc[,prob.prod := FINEMAP*PIP,]
		setorder(df_coloc, Trait, -prob.prod)
		ymax = max(df_coloc[,prob.prod])
		fig_coloc = ggplot(df_coloc[prob.prod>0,], aes(Position, FINEMAP*PIP, color=Trait, size=prob.prod^2)) + geom_point() + scale_x_continuous(expand=c(0,0)) + ylab("CLPP") + scale_size_continuous(limits = c(0, 1), range(0.1, 6)) + ylim(0,1) + scale_color_manual(values = cols) + geom_text_repel(data=subset(df_coloc[prob.prod>0,], !duplicated(Trait)), aes(label=Trait), size=3, force=1, nudge_y = .1, nudge_x=window/2, hjust=0)
	}else{	
		fig_coloc = ggplot() + scale_x_continuous(expand=c(0,0), limits=c(start(wh), end(wh))) + ylim(0,1) + ylab("CLPP")
	}

	# Gene models     
	fig_genebody = plotEnsGenes_gg( EnsDb.Hsapiens.v75, start(wh), end(wh), seqnames(wh), splice_variants=FALSE, non_coding=non_coding, ensGene=ensGene) 
	
	# Combine plots
	fig_track = ggbio::tracks( "eQTL"		= fig_eqtl, 
		"eQTL\nFine map"	= fig_finemap_gene,
		"GWAS\nFine map" 	= fig_causal,
		"Shared\ncandidates"= fig_coloc,
		"Genes"				= fig_genebody,
		xlim = wh,
		padding = unit(-.65, "lines"),
		label.bg.fill="navy", label.text.color="white",
		heights=c(.9,.5,.5,.5,.4),
		theme = theme_bw(8) + theme(legend.position="none", panel.grid.minor = element_blank()),
			title=paste0( geneSymbol, ' (',ensGene,')') )
	fig_track@mutable['Genes'] = FALSE
	fig_track
}


  # tracks(fig_eqtl, fig_finemap_gene, xlim=wh)









